ci: use github action for automated deployment

ttshivers opened the pull request:
This replaces the Travis ci automated deployment with a GitHub action.

Requires the use of some bot token to make a commits at [nodejs-github-bot/official-images](https://github.com/nodejs-github-bot/official-images) and then open PRs on [docker-library/official-images](https://github.com/docker-library/official-images) .

Refs: #1312

tianon commented on .github/prFormat.js:
```diff
@@ -0,0 +1,43 @@
+  // Use pr title if available or fallback to first line of head commit message
+  const title = total_count
+    ? pr.title
+    : payload.head_commit.message.split('\n')[0];
```
Given this is being triggered by pushes to `master` (not pull requests), won't it never have a PR object?  (and in that case, the `head_commit` is going to be the merge commit, and thus generate a problematic title/commit message)

ttshivers replied:
I do a search for PRs with the sha of the head_commit that was pushed.  It's a similar approach to how [semantic-release](https://github.com/semantic-release/semantic-release/blob/master/index.js) looks for issues/PRs

nschonni commented:
Cool, beat me to it. Here is what I was working on locally, because there are probably some bits you might want

```yaml
name: PR to official images

on:
  pull_request_target:
    types:
      - closed
    paths:
      - "**/Dockerfile"
      - "**/architectures"
      - "**/docker-entrypoint.sh"
      - "generate-stackbrew-library.sh"
      - "functions.sh"

jobs:
  pr:
    runs-on: ubuntu-latest
    # if: github.repository_owner == 'nodejs'

    steps:
      - name: Checkout the docker-node repo
        uses: actions/checkout@v2
        with:
          path: docker-node
          fetch-depth: 50

      - name: Checkout the docker-library/official-images repo
        uses: actions/checkout@v2
        with:
          path: official-images
          repository: nschonni/official-images

      - name: Generate Stackbrew for diff
        run: |
          cd docker-node
          ./generate-stackbrew-library.sh > ../official-images/library/node

      - uses: peter-evans/create-pull-request@v3
        with:
          token: ${{ secrets.GITHUB_API_TOKEN }}
          # push-to-fork: nodejs-github-bot/official-images
          title: "Node: ${{ github.event.pull_request.title}}"
          path: official-images
          commit-message: "Pull Request: {{ github.event.pull_request.html_url }}"
```

nschonni commented on .github/workflows/auto_pr.yml:
```diff
@@ -0,0 +1,82 @@
+          script: |
+            const basePath = `${process.env.GITHUB_WORKSPACE}/docker-node`;
+            const script = require(`${basePath}/.github/prFormat.js`);
+            return script({ github, context, maintainersFile: `${basePath}/ACTIVE_MAINTAINERS` });
```
Think we could even just drop the ACTIVE_MAINTAINERS and just hardcode the list. AFAIK, it's only used for the PR flow

nschonni commented on .github/workflows/auto_pr.yml:
```diff
@@ -0,0 +1,82 @@
+        uses: peter-evans/create-pull-request@v3
+        with:
+          token: ${{ secrets.MACHINE_USER_PAT }}
+          push-to-fork: ${{ secrets.MACHINE_USER_FORKED_REPOSITORY }}
```
Doesn't need to be a secret, just "docker-library/official-images" should be fine. It's a public fork https://github.com/nodejs-github-bot/official-images

ttshivers commented:
I have extra complexity in mine that handles if commits are pushed to master without going through a pull request. Should I assume that won't happen and take out some of that stuff?

nschonni reviewed (commented):
Few of the things you're grabbing through scripts are available in the github.event.pull_request object.
That's available in the pull_request_target version, and has the benifit of the built in repo token working for the comment part (but not the PR/push)

nschonni commented:
The main branch is protected, so it can only be changed through PRs. That should simplify some stuff for you 😄

ttshivers commented:
@nschonni I see in your approach, you are using the `pull_request_target` trigger. It seems like it does have some benefits in that you have access to the PR without having to search for it. I changed over to that method and added the other differences too.

nschonni commented on .github/workflows/official-pr.yml:
```diff
@@ -0,0 +1,63 @@
+          push-to-fork: nodejs-github-bot/official-images
+          path: official-images
+          branch: autopr
+          branch-suffix: short-commit-hash
```
```suggestion
          branch-suffix: ${{ github.event.pull_request.number }}
```

ttshivers replied:
The suffix field is more of an enum and behaves this way: https://github.com/peter-evans/create-pull-request/blob/da928d5fcce0b0dd0c34642f77e862dc9cf76d8b/dist/index.js#L326

However, we can just remove the `branch-suffix` option and add the pr number to `branch`

nschonni replied:
Ah, OK. Not a big one, it's fine as-is

nschonni commented on .github/workflows/official-pr.yml:
```diff
@@ -0,0 +1,63 @@
+          commit-message: "Node: ${{ github.event.pull_request.title }}"
+          title: "Node: ${{ github.event.pull_request.title }}"
+          body: |
+            ${{ github.event.repository.full_name }}#${{ github.event.pull_request.number }}
```
```suggestion
            Pull Request: ${{ github.event.pull_request.html_url }}
```

nschonni reviewed (commented):
Looking good!

nschonni commented on .github/workflows/official-pr.yml:
```diff
@@ -0,0 +1,63 @@
+
+      - name: Create PR comment
+        uses: peter-evans/create-or-update-comment@v1
+        with:
```
Probably need `github-token: ${{secrets.GITHUB_TOKEN}}`

nschonni replied:
Nevermind, I was thinking of another action, looks like this one already assumes this

nschonni commented:
@nodejs/docker (or maybe specifically @LaurentGoderre) does anyone have the token that we used for Travis that they could put in the GitHub secrets section?

SimenB reviewed (approved):
this makes me very happy 😀 Perfect #1337 👍

SimenB commented:
@nodejs/build-infra could you add the bot's token to the repo's secrets? Or is that stuff managed by some other group?

SimenB commented on .github/workflows/official-pr.yml:
```diff
@@ -0,0 +1,63 @@
+          ./generate-stackbrew-library.sh > ../official-images/library/node
+
+      - name: Create PR in official-images
+        id: cpr
```
full name instead of some abbreviation?

ttshivers replied:
How's `create-pr`

nschonni commented:
Created https://github.com/nodejs/admin/issues/556 since it seems like there is more of an official process now around the tokens and secrets

rvagg commented:
docker_node_github_api_token.md is in build/github-bot in the secrets repo already, I assume that's what you're asking for?

nschonni commented:
@rvagg yup, that would work! We're just looking to switch the CI the PR setup is running on

mmarchini commented:
Still need approval on the admin repo. I can setup the secrets on this repo when request is approved.

mmarchini commented on .github/workflows/official-pr.yml:
```diff
@@ -0,0 +1,63 @@
+        id: create-pr
+        uses: peter-evans/create-pull-request@v3
+        with:
+          token: ${{ secrets.GITHUB_API_TOKEN }}
```
I'm 90% sure `secrets.GITHUB_TOKEN` will work here.

ttshivers replied:
I am not completely sure. The user needs to be able to push commit to some branch on `nodejs-github-bot/official-images` and open pull requests in `docker-library/official-images`. 

https://docs.github.com/en/actions/reference/authentication-in-a-workflow#permissions-for-the-github_token

mmarchini replied:
Yeah it definitely doesn't have permission for the first one. If it was a branch in this repo it could potentially work.

mmarchini commented on .github/workflows/official-pr.yml:
```diff
@@ -0,0 +1,63 @@
+name: Create official images PR
+
+on:
+  pull_request_target:
+    types:
+      - closed
```
Is `pull_request_target` the best event for this? Why is it better than `push`?

ttshivers replied:
I originally had `push` and did an API request to fetch the associated PRs, but I switched to `pull_request_target` because the context included the PR number.

mmarchini replied:
I see. The downside is that any of the 15 other PRs open right now won't trigger this event unless they are rebased before merge. It also doesn't trigger if someone pushes directly to master.

I'm wondering if: a) we can replace `github.event.pull_request.title` with the commit message; and b) for pushes coming from PRs, if `github.event.pull_request` is available (would need to test it in a separate repo).

Not blocking and I'm not familiar with this repo commit workflow, so @nodejs/docker folks will know better which event to use. But `pull_request_target` has drawbacks, just pointing those so the team makes the right decision.

ttshivers replied:
If this is merged before the other PRs, I am thinking it will still be triggered because `master` will have this action and should detect a PR being closed. I am unsure of these semantics. I will run a quick test to see what happens.

mmarchini replied:
ah right, _maybe_ it works. `pull_request_target` triggers are kinda meh though, so I wouldn't be surprised if it didn't work.

ttshivers replied:
Alright ran the test and it does trigger as long as it's in `master`. However, I agree with you and would prefer it being on `push` to `master` instead too. I'd prefer it to be a `push` trigger too. I think I could replicate the behavior with a version shorter than my original one by adding a step like:
```
      - uses: actions/github-script@v3
        id: pr
        with:
          result-encoding: string
          script: |
            const result = await github.repos.listPullRequestsAssociatedWithCommit({
              owner: context.payload.repository.owner.name,
              repo: context.payload.repository.name,
              commit_sha: context.payload.head_commit.id
            })
            return result.data[0].number;
```

> The main branch is protected, so it can only be changed through PRs. That should simplify some stuff for you 😄

This also gives me confidence that I wouldn't have to worry about pushes to `master` happening through anything else other than from a PR.

mmarchini replied:
Thanks for testing it. Yes, I missed the master branch is protected here, so we should be good :)

SimenB commented:
One more question, is it possible to _not_ run this if tests or building an image fail? Or is the thinking that since this only runs on master (which is protected), CI was green before merging so we don't need to wait.

Looking at https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#jobsjob_idneeds it doesn't seem like the ID is referencable across workflows. And we wouldn't know which to wait for either as builds are triggered conditionally. I guess we could try to have one workflow instead of many? Not sure how that would look

mmarchini commented:
The deployment steps would need to be on the same job as the tests so if a test fails this job doesn't run, or the job has to be in the same file and can depend on the test/build job.

ttshivers commented:
> The deployment steps would need to be on the same job as the tests so if a test fails this job doesn't run, or the job has to be in the same file and can depend on the test/build job.

I like this way too. Another way is to use some an action like [lewagon/wait-on-check-action](https://github.com/lewagon/wait-on-check-action) to wait for a test workflow to finish before allowing another one to run.

nschonni commented:
@rvagg we got the thumbs up over on the admin repo https://github.com/nodejs/admin/issues/556#issuecomment-700458554 if you have access to set the secret

rvagg commented:
no, I don't think I have all the required access, @mmarchini might be the best person for this?

nschonni commented:
From @yosifkit https://github.com/docker-library/official-images/pull/8824#issuecomment-702955542

> I think we'd rather that the bot be improved to only open a single PR at a time.

Maybe we should drop the branch suffix and just keep a single "official-images" branch that it could keep pushing to the same PR till landed

nschonni commented on .github/workflows/official-pr.yml:
```diff
@@ -0,0 +1,62 @@
+        id: create-pr
+        uses: peter-evans/create-pull-request@v3
+        with:
+          token: ${{ secrets.GITHUB_API_TOKEN }}
```
There was a small change in the name due to a GitHub naming issue
```suggestion
          token: ${{ secrets.GH_API_TOKEN}}
```
Feel free to just rebase this into your commit

nschonni commented:
I think we're good to land this, just wanted to walk through the flow now that we're re-using the branch

Flow 1:
1. PR A lands here
1. Action creates upstream PR ##
1. Action comments on PR A
1. Upstream PR lands. Branch is leftover or deleted?

Flow 2:
1. PR A lands here
1. Action creates upstream PR ##
1. Action comments on PR A
1. PR B lands here
1. Action pushes new commits. Title stays the same?
1. Action comments on PR B
1. Upstream PR lands. Branch is leftover or deleted?

Flow 3:
1. PR A lands here
1. Action creates upstream PR ##
1. Action comments on PR A
1. Upstream PR is rejected/closed. Branch is leftover or deleted?

ttshivers commented:
The action uses `peter-evans/create-pull-request`, and they describe their behavior like: https://github.com/peter-evans/create-pull-request#action-behaviour.

In Flow 1, currently the branch is NOT deleted because `delete-branch` is not specified and defaults to false. This can be changed if you want.

In Flow 2, step 5, the PR title will be updated (so will the PR body). Step 7, branch is currently not deleted because of what I mentioned above.

In Flow 3, branch is not deleted for same reason.

I think it may be a good idea to set `delete-branch` to true because it will cause the branch to be deleted in all 3 flows, which seems desirable.

nschonni commented:
I can't tell from the docs if there would need to be some sort of `repository_dispatch` setup on the other side for the delete branch to work, so I'm good leaving it for now.
I think there maybe be an issue with the fact we're using the base action/checkout then using the PR action to re-use/update the branch, but unless anyone has any last minute changes, I think we should land this and then test it out by landing #1347 then #1348

ttshivers commented:
> I can't tell from the docs if there would need to be some sort of `repository_dispatch` setup on the other side for the delete branch to work, so I'm good leaving it for now.
> I think there maybe be an issue with the fact we're using the base action/checkout then using the PR action to re-use/update the branch, but unless anyone has any last minute changes, I think we should land this and then test it out by landing #1347 then #1348

It appears that the action deletes the branch the next time it is invoked on this repository (and there is no diff). https://github.com/peter-evans/create-pull-request/blob/e17bb55cb71e482dbeb01f204bff23f21dd20c77/src/create-pull-request.ts#L204

nschonni commented:
OK, I'm going to land this and the other 2 and we'll see how it goes!
Thanks for all the work

nschonni commented:
@mmarchini FYI, we did run into the merge issue with the first PR using the `pull_request_target`, so I guess the regular merge isn't the workaround and we'd need to switch the strategy to rebase merge

ttshivers commented:
> @mmarchini FYI, we did run into the merge issue with the first PR using the `pull_request_target`, so I guess the regular merge isn't the workaround and we'd need to switch the strategy to rebase merge

Yeah I messed up in my initial testing of the trigger. When I run the tests again, its behaving the same way it's behaving in this repo now.
